### PR TITLE
[TwigBundle] do not try to register incomplete definitions

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20212
| License       | MIT
| Doc PR        |

### CODE REVIEW

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php [hunk 1]
```diff
@@ -151,4 +161,36 @@ public function getNamespace()
     {
         return 'http://symfony.com/schema/dic/twig';
     }
+
+    private function cleanUpDefinitionsMissingRequiredDependencies(ContainerBuilder $container)
+    {
+        if (!interface_exists('Symfony\Component\Templating\TemplateReferenceInterface')) {
+            $container->removeDefinition('twig.controller.exception');
+        }
```
Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: This one looks wrong to me as the Exception controller only depends on Twig, not the Templating component.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php [hunk 2]
```diff
@@ -151,4 +161,36 @@ public function getNamespace()
...
+
+        if (!class_exists('Symfony\Component\Debug\Exception\FlattenException') && !interface_exists('Symfony\Component\EventDispatcher\EventSubscriberInterface')) {
```
Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: Shoudn't it be a `||` instead of `&&`

**Comment by fabpot** [comment]
When merging this PR in 3.2, we should add checks for each `if` with the new `ClassExistenceResource` resource.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php (hunk 1 above)
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: `findTemplate()` returns a `TemplateReferenceInterface` instance: https://github.com/symfony/symfony/blob/2.7/src/Symfony/Bundle/TwigBundle/Controller/ExceptionController.php#L99

Technically, this is not needed in 2.7 as the FrameworkBundle already requires the Templating component. But this check IMO is more correct as we do not rely on FrameworkBundle internals here.

**Comment by xabbuh** [comment]
ping @symfony/deciders :)

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php (hunk 1 above)
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: This should be removed in newer versions where a string is returned though.

File: src/Symfony/Bundle/TwigBundle/Resources/config/form.xml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" ?>
+<container xmlns="http://symfony.com/schema/dic/services"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <parameters>
+        <parameter key="twig.extension.form.class">Symfony\Bridge\Twig\Extension\FormExtension</parameter>
+        <parameter key="twig.form.engine.class">Symfony\Bridge\Twig\Form\TwigRendererEngine</parameter>
+        <parameter key="twig.form.renderer.class">Symfony\Bridge\Twig\Form\TwigRenderer</parameter>
```
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/Resources/config/form.xml]: This parameters may know not be defined while they are before, which can be a BC break. So not god for a 2.7.x release.

**Comment by stof** [comment]
And regarding the extensions, we already have checks in a compiler pass. So I would rather fix them here if they are missing some condition rather than  doing a second system in the DI extension itself.

File: src/Symfony/Bundle/TwigBundle/Resources/config/form.xml (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/Resources/config/form.xml]: I moved them back to the `twig.xml` file.

**Comment by xabbuh** [comment]
I also moved most of the clean up logic into the compiler passes except for the `twig.translation.extractor` service which didn't fit well in any of the existing passes.

**Comment by HeahDude** [comment]
This one should be good to merge now.

👍

Status: Reviewed

**Comment by fabpot** [comment]
Thank you @xabbuh.

**Pull request merged**